lib/rucio/core/did_meta_plugins/json_meta.py

```diff
@@ -150,41 +139,77 @@ def delete_metadata(self, scope, name, key, session=None):
 
     @stream_session
     def list_dids(self, scope, filters, did_type='collection', ignore_case=False, limit=None,
-                  offset=None, long=False, recursive=False, session=None):
-        # Currently for sqlite only add, get and delete is implemented.
+                  offset=None, long=False, recursive=False, ignore_dids=None, session=None):
         if not json_implemented(session=session):
             raise NotImplementedError
 
-        query = session.query(models.DidMeta)
-        if scope is not None:
-            query = query.filter(models.DidMeta.scope == scope)
-        filters.pop('name', None)
-        for k, v in iteritems(filters):
-            if session.bind.dialect.name == 'oracle':
-                query = query.filter(text("json_exists(meta,'$?(@.{} == \"{}\")')".format(k, v)))
-            else:
-                query = query.filter(cast(models.DidMeta.meta[k], String) == type_coerce(v, JSON))
-
-        if long:
-            for row in query.yield_per(5):
-                yield {
-                    'scope': row.scope,
-                    'name': row.name,
-                    'did_type': 'Info not available in JSON Plugin',
-                    'bytes': 'Info not available in JSON Plugin',
-                    'length': 'Info not available in JSON Plugin'
-                }
-        else:
-            for row in query.yield_per(5):
-                yield row.name
+        if not ignore_dids:
+            ignore_dids = set()
+
+        # backwards compatability for filters as single {}.
+        if isinstance(filters, dict):
+            filters = [filters]
+
+        # instantiate fe and create sqla query, note that coercion to a model keyword
+        # is not appropriate here as the filter words are stored in a single json column.
+        fe = FilterEngine(filters, model_class=models.DidMeta, strict_coerce=False)
+        query = fe.create_sqla_query(
+            additional_model_attributes=[
+                models.DidMeta.scope,
+                models.DidMeta.name
+            ], additional_filters=[
+                (models.DidMeta.scope, operator.eq, scope)
+            ],
+            json_column=models.DidMeta.meta
+        )
+
+        if limit:
+            query = query.limit(limit)
+        if recursive:
+            from rucio.core.did import list_content
```

Comment:
This can't be loaded at top? Circular dependency?

Reply:
Not sure of the reasoning (someone else did this). Doesn't look circular but get a bunch of import errors when it's at the top.